File: code/tinyos/apps/TxIndexCoding/multirandperm.py
import numpy as np

def multirandperm(K, n):
	inds = np.zeros((n,K))
	for i in range(n):
		inds[i,:] = singlerandperm(K)
	return inds

def singlerandperm(K):
	perm = np.random.permutation(K)
	rangeinds = np.arange(K)
	reorder = rangeinds[perm == rangeinds]
	for i in range(0, len(reorder), 2):
		if i == len(reorder) - 1:
			ind = i % (K-1)
			perm[reorder[i]] = perm[ind]
			perm[ind] = reorder[i]
		else:
			# switch the two elements
			perm[reorder[i]] = reorder[i+1]
			perm[reorder[i+1]] = reorder[i]
	return perm

def pairingperm(K):
	perm = []
	for i in range(0, K-1, 2):
		perm.extend(range(i, i+2)[::-1]);
	if K % 2 == 1:
		perm[0] = K-1
		perm.append(1)
	return np.array(perm)

File: code/matlab/multirandperm.m
function [inds] = multirandperm(K, n, all)
%multirandperm "random" permutation
% For each element (receiver), assign n indices (messages) that differ from the element
% Inputs:
%  K - number of elements
%  n - number of indices per element
%  all - true/false specifying whether each element is used as an index
%   (i.e. without replacement)
% Outputs:
%  inds [n-by-K]


inds = zeros(n,K);
for i = 1:n
    inds(i,:) = singlerandperm(K);
end
